lemmatize_text_data_WIP.py

```diff
@@ -0,0 +1,160 @@
+from pathlib import Path
+import pandas as pd
+from preprocess_data import change_column_types, split_train_val_test
+import requests
+import json
+from time import time
+
+
+def get_lemmas(text: str) -> list:
+    """
+    for a given Czech text returns a list of lemmas
+    calls NLP FI MUNI tagger api
+    """
+    data = {"call": "tagger",
+            "lang": "cs",
+            "output": "json",
+            "text": text.replace(';', ',')     # very important!!! if semicolon not replaced by comma, json can't read it
+            }
+    uri = "https://nlp.fi.muni.cz/languageservices/service.py"
+    response = requests.post(uri, data=data, timeout=10000)
+    response.raise_for_status()  # raises exception when not a 2xx response
+    byte_data = response.content
+    data = json.loads(str(byte_data, 'utf-8'))
+    tokens = [token for token in data["vertical"] if len(token) == 3]
+    lemmas = []
+
+    for token in tokens:
+        lemmas.append(token[1])
+    return lemmas
+
+
+def get_lemmatized_column(column: pd.Series) -> pd.Series:
+    """
+    Returns the same pd.Series but with text as lemmas inside.
+    Somewhat wrapper for get_lemmas function because NLP FI MUNI API does not allow more than 500 calls/day
+    so the column has to be joint in one text and then split again and not send a cell by cell
+    """
+    joint_rows = ""
+    for row in column:
+        joint_rows += row + " XXX "
+
+    lemmas = get_lemmas(joint_rows)
+
+    lemmatized_row = ""
+    lemmatized_rows = []
+    for lemma in lemmas:
+        if lemma != "XXX":
+            lemmatized_row += lemma + " "
+        else:
+            lemmatized_rows.append(lemmatized_row)
+            lemmatized_row = ""
+
+    return pd.Series(lemmatized_rows)
+
+
+def get_lemmatized_df(dataframe: pd.DataFrame) -> None:
+    """
+    creates a dataframe with lemmatized texts
+    takes long time to execute!!! (around 20 minutes) because of the calls to external API with all text data
+    it is split into so many parts because we want to avoid the HTTP run out of the time exception
+    """
```

Comment:
it is better to use lemmatization library instead of API call (isn't Prague lemmatizer free?)
it is good to cache results not to need to run it again and again
e.g. make JSON with already translated sentences

Reply:
I suggest to use pymorphodita, they have better models and tools than NLP lab

Not sure if it works, but there is a chatGPT example. I suggest to make lemmatizer module and import it as needed:

```
import pymorphodita

# Load the pre-trained Czech model
def load_morphodita_model(model_path):
    tagger = pymorphodita.Tagger.load(model_path)
    if not tagger:
        raise Exception("Failed to load the MorphoDiTa model")
    return tagger

# Lemmatize the input Czech text
def lemmatize_text(tagger, text):
    tokenizer = pymorphodita.Tokenizer.new_utf8_tokenizer()
    tokenizer.set_text(text)
    
    lemmas = []
    forms = pymorphodita.Forms()
    lemmas_list = pymorphodita.TaggedLemmas()
    
    while tokenizer.next_sentence(forms):
        tagger.tag(forms, lemmas_list)
        for lemma in lemmas_list:
            lemmas.append(lemma.lemma.split('_')[0])  # Extracting the base lemma
    return lemmas

# Example usage
if __name__ == "__main__":
    model_path = 'czech-morfflex-pdt-161115.tagger'  # Path to the MorphoDiTa model
    tagger = load_morphodita_model(model_path)
    
    text = "Byl jsem včera na návštěvě u babičky."  # Example Czech sentence
    lemmas = lemmatize_text(tagger, text)
    
    print("Lemmas:", lemmas)
```

Reply:
the code above is crap, just example of morphodita using
It uses line comments instead of docstrings, it is not using default value for model path, etc.

Reply:
I wasn't able to find/install pymorphodita specifically but I ahve used ufal.morphodita which I believe should be basically the same.